### PR TITLE
Changed the client so that GTIDs are updated before calling the event listeners

### LINKED ISSUE #122: BinaryLogClient calls event listeners before updating client's position
The BinaryLogClient [processes an event](https://github.com/shyiko/mysql-binlog-connector-java/blob/fb379adeec1648249b25705a6d08974db5533db4/src/main/java/com/github/shyiko/mysql/binlog/BinaryLogClient.java#L748) by first calling the registered event listeners for the event type, and _then_ updating the client's position (i.g., binlog filename & position, GTID set) for that event.

The code may have done it this way because each event header contains the binlog filename and position _of the next event_. This means that because the event handlers are called before the client's binlog filename and position are updated (to the next event), when an event handler is processing an event the client's binlog and filename happen to point to the binlog filename and position _of this event_. 

But GTIDs don't work this way. The `GTID` event that precedes our event of interest has the GTID for _subsequent events_, unlike the event header's binlog filename and position that reflect the _next_ event's position. By updating the client's GTID set _after_ the event listeners are called for the GTID event means that any event listeners reading the client's GTID set see the _GTID set_ before this new transaction, and thus are incorrect.

Fortunately, this behavior affects only the event handlers for `GTID` events.

### PR DESCRIPTION
**Pull request by rhauch**
Previously, GTIDs were updated at the same time the binlog filename and position were updated: after the event listeners were called. However, this means that the GTID set was not yet updated when the event listeners were called, meaning that the event listeners would see the old GTID set if they asked for it.

With this change, the GTID set is updated before the event listeners are called, and the binlog filename and position (which reference the _next_ event) are updated after the event listeners are called. This ensures that the event listeners see the event's correct GTID set and binlog filename and position.

fixes #122

### CODE REVIEW

File: src/main/java/com/github/shyiko/mysql/binlog/BinaryLogClient.java [hunk 1]
```diff
@@ -779,41 +782,52 @@ private void listenForEventPackets() throws IOException {
...
+                    if (advanceGTID()) {
+                        if (logger.isLoggable(Level.WARNING)) {
+                            logger.log(Level.WARNING, "GtidSet wasn't synchronized before GTID. " +
+                                    "Please submit a bug report to https://github.com/shyiko/mysql-binlog-connector-java");
+                        }
                     }
                 }
+                return true;
             }
-            previousGtidEvent = event;
```
Comment (rhauch) [inline src/main/java/com/github/shyiko/mysql/binlog/BinaryLogClient.java]: The github diff makes this difficult to see, but this is in the `updateGtid(Event)` method. The line does appear in the `updatePosition(Event,boolean)` method [here](https://github.com/shyiko/mysql-binlog-connector-java/pull/125/files#diff-7addc4e3eed6e9254e088abc015ac8adR837).

Comment (shyiko) [inline src/main/java/com/github/shyiko/mysql/binlog/BinaryLogClient.java]: "previousGtidEvent" != "previousEvent" 🐱

Comment (rhauch) [inline src/main/java/com/github/shyiko/mysql/binlog/BinaryLogClient.java]: ah, gotcha.

**Comment by shyiko** [comment]
@rhauch Thank you! 🙇

**Pull request merged**